### PR TITLE
SL-667: Add CICD config for Javabuilder demo env

### PR DESCRIPTION
**Pull request by ebeastlake**
Adds a new `javabuilder-demo` stack for Java Lab eval mode.

Tickets:
- https://codedotorg.atlassian.net/browse/SL-666
- https://codedotorg.atlassian.net/browse/SL-667

Outstanding questions:
- [x] ~~How do we modify `javabuilder/template.yml.erb` so that it doesn't use (or effectively ignores) the `TeacherAssociatedRequests` table ([here](https://github.com/code-dot-org/javabuilder/blob/main/cicd/3-app/javabuilder/template.yml.erb#L943)) and the `TeacherLimitPerHour` ([here](https://github.com/code-dot-org/javabuilder/blob/main/cicd/3-app/javabuilder/template.yml.erb#L394))?~~ Will address this later per https://github.com/code-dot-org/javabuilder/pull/385#discussion_r1160054241.

### CODE REVIEW

File: cicd/3-app/javabuilder/config/production-demo.config.json
Comment (cat5inthecradle) [reply cicd/3-app/javabuilder/config/production-demo.config.json:11]: But this would be safe to deploy as-is. You could also remove that config parameter at this time, and add it later as it's own PR.

Comment (cat5inthecradle) [reply cicd/3-app/javabuilder/config/production-demo.config.json:11]: In fact, that's what I'd recommend. Remove the TeacherLimitPerHour part from this PR, then we can merge this and make sure we've got a green pipeline into this new environment before adding new config parameters.

**Review by cat5inthecradle — APPROVED** [review]
LGTM! To deploy this change, after merging, we just need to run the following.

```
cicd/2-cicd/deploy-cicd.sh
```

There might be a bit of pain to do with ruby versions or AWS credentials. Once you merge, switch back to `main` on your local, pull this new code, then run `cicd/2-cicd/deploy-cicd.sh`. Ping me on slack and I can follow along and help troubleshoot any issues.

**Comment by sanchitmalhotra126** [comment]
> I specified the `javabuilder-demo` subdomain in the YML file config, so this template should produce `wss://javabuilder-demo.code.org`, but where does that value get exposed? (i.e. These templates and URLs are 1-1 with the env files, but I need dashboard to be 1-2 with URLs, and I don't see where I do something like `JavabuilderDemoURL = *`).

Yep, updating Dashboard (and the frontend) would be a separate change - 

Dashboard reads the Javabuilder websocket and HTTP endpoints [here](https://github.com/code-dot-org/code-dot-org/blob/b9c723e31d9436038c37afcffa2b6efa970a1405/lib/cdo.rb#L142) and [here](https://github.com/code-dot-org/code-dot-org/blob/b9c723e31d9436038c37afcffa2b6efa970a1405/lib/cdo.rb#L161) (in `cdo.rb`). Currently we just default to the prod URL (unless overridden via DCDO, or by local environment config), so we'll have to update some logic to serve the correct URL to the client based on the situation. We read this value from the config in our back end [here](https://github.com/code-dot-org/code-dot-org/blob/b9c723e31d9436038c37afcffa2b6efa970a1405/dashboard/app/models/levels/javalab.rb#L131) in javalab.rb, which is served the front end. The actual websocket request, using the given URL, is made [here](https://github.com/code-dot-org/code-dot-org/blob/b9c723e31d9436038c37afcffa2b6efa970a1405/apps/src/javalab/JavabuilderConnection.js#L174) (in JavabuilderConnection). The HTTP url is used in the backend only when making the initial put sources call; that's read [here](https://github.com/code-dot-org/code-dot-org/blob/b9c723e31d9436038c37afcffa2b6efa970a1405/dashboard/app/helpers/javalab_files_helper.rb#L3) (in javalab_files_helper).

File: cicd/2-cicd/cicd.template.yml
Comment (sanchitmalhotra126) [inline cicd/2-cicd/cicd.template.yml:409]: Probably a question for @cat5inthecradle - is it possible/should it be possible to target a branch other than `main` for a production deploy?

**Review by sanchitmalhotra126 — APPROVED** [review]
One question out of curiosity, but otherwise LGTM!

**Pull request merged**
